Tensile/Source/client/source/PerformanceReporter.cpp

```diff
@@ -0,0 +1,178 @@
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+
+#include "PerformanceReporter.hpp"
+
+#include <cmath>
+#include <cstddef>
+#include <limits>
+#include <string>
+#include <unordered_map>
+
+namespace Tensile
+{
+    namespace Client
+    {
+        std::shared_ptr<PerformanceReporter> PerformanceReporter::Default(po::variables_map const& args)
+        {
+            int     deviceIndex = args["device-idx"].as<int>();
+            double  l2ReadHits = args["perf-l2-read-hits"].as<double>();
+            double  l2WriteHits = args["perf-l2-write-hits"].as<double>();
+            double  readEff = args["perf-read-efficiency"].as<double>();
+            bool    mfma = args["perf-mfma"].as<bool>();
+ 
+            return std::make_shared<PerformanceReporter>(deviceIndex, l2ReadHits, l2WriteHits, readEff, mfma);
+        }
+        
+        PerformanceReporter::PerformanceReporter(int deviceIndex, double l2ReadHits, double l2WriteHits, double readEff, bool mfma)
+        {
+            hipGetDeviceProperties(&m_props, deviceIndex);
+            setNumCUs();
+            setMemoryBusWidth();
+            setPerfModel(l2ReadHits, l2WriteHits, readEff, mfma);
+            setMagicNum();
+            m_deviceProps = true;
+            
+            perf.l2ReadHitRate = getL2ReadHits();
+            perf.l2WriteHitRate = getL2WriteHits();
+            perf.readEff = getReadEff();
+            perf.CUs = getNumCUs();
+        }
+        
+        void PerformanceReporter::reportValue_uint(std::string const& key, uint64_t value) 
+        {
+            if(key == ResultKey::SpeedGFlops && m_deviceProps) 
+            {
+                reportValue_numeric(key, value);
+            }
+        }
+
+        void PerformanceReporter::reportValue_double(std::string const& key, double value) 
+        {
+            if(key == ResultKey::ClockRateSys && m_deviceProps)
```

Comment:
reportValue_double seems an odd place to if-check the key and record clocks in member vars like m_memClockMhz- should be done in a constructor or elsewhere?